returnn/hdf.py

```diff
@@ -346,6 +366,278 @@ def run(self):
+            start_and_duration will round down the start time and the duration of the segment
+            rasr_compatible will round up the start time and round down the end time
+        :param round_factor: do the rounding based on a sampling rate that is scaled down by this factor
+        :param target_sampling_rate: desired sampling rate for the HDF, data will be resampled to this rate if needed
+        :param skip_on_error: If True, skip segments that fail to be processed (for known reasons only) instead
+            of failing the job.
+        """
+        self.bliss_corpus = bliss_corpus
+        self.splits = splits
+        assert concurrent > 0
+        self.concurrent = concurrent
+        if compress_format is not None:
+            assert compress_factor is not None
+            assert 0 <= compress_factor
+            assert output_dtype is None
+            self.output_dtype = "int32"
+        else:
+            assert output_dtype in ["float64", "float32", "int32", "int16"]
+            self.output_dtype = output_dtype
+        self.compress_format = compress_format
+        self.compress_factor = compress_factor
+        self.multi_channel_strategy = multi_channel_strategy or BlissToPcmHDFJob.Mixdown()
+        assert isinstance(self.multi_channel_strategy, (BlissToPcmHDFJob.Mixdown, BlissToPcmHDFJob.PickNth))
+        self.rounding = rounding
+        assert round_factor > 0
+        self.round_factor = round_factor
+        self.returnn_root = returnn_root
+        self.target_sampling_rate = target_sampling_rate
+        self.skip_on_error = skip_on_error
+
+        self.out_hdfs = [self.output_path(f"{i + 1:0d}.hdf") for i in range(len(splits))]
+
+        num_cores = min(len(splits) + 1, 16)
+        mem_per_core = 1
+        self.rqmt = {"cpu": num_cores, "mem": num_cores * mem_per_core + 2, "time": 48}
+
+    def tasks(self):
+        yield Task("run", rqmt=self.rqmt, args=range(self.concurrent))
+
+    def run(self, index: int):
+        out_hdfs = list(util.chunks(self.out_hdfs, self.concurrent))[index]
+        splits = list(util.chunks(self.splits, self.concurrent))[index]
+        assert len(out_hdfs) == len(splits)
+        _logging.info(f"Writing audio to {out_hdfs} from {splits}.")
+
+        SimpleHDFWriter = get_returnn_simple_hdf_writer(self.returnn_root.get_path())
+        hdf_writers = [SimpleHDFWriter(filename=out_hdf.get_path(), dim=1) for out_hdf in out_hdfs]
+        segment_whitelists = []
+        for split in splits:
+            with uopen(split, "rt") as f:
+                segments_whitelist = {line.strip() for line in f if len(line.strip()) > 0}
+            segment_whitelists.append(segments_whitelist)
+
+        assert len(segment_whitelists) == len(out_hdfs)
+        all_whitelists = set.union(*segment_whitelists)
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        pool = multiprocessing.Pool(max(self.rqmt["cpu"] - 1, 1))
```

Comment:
I am wondering why we have two levels of parallelism: 
* run `concurrent` many sisyphus tasks on possibly different nodes
* run `self.rqmt["cpu"]-1` many workers for multiprocessing

could we achieve similar speed by just using one form of parallelism?

Reply:
There are two levels here because with every unit of `concurrent` I/O becomes less efficient, as we can end up reading a single audio file multiple times.

Within one concurrency unit the code avoids reading the audio sources more than once and is therefore as I/O efficient as can be. As the corpora grow, however, the single job instance (with multiprocessing) becomes too little and then you can eventually start to tolerate some I/O inefficiency.